Add ITK-VTK-Viewer example

PaulHax opened the pull request:
closes #3 
closes #7

PaulHax committed:
chore(formating): add prettier and run it

PaulHax committed:
chore(control points): convert to classes

PaulHax committed:
fix(tests): update vitest version to fix EventTarget triggering tests

PaulHax committed:
chore(actions): lint on ci action

PaulHax committed:
feat(Container): cleanup added svg element with remove

PaulHax committed:
feat(PointsContoller): clicking adds a new control point

PaulHax committed:
chore(prettier): check styling during CI

PaulHax committed:
feat(control points): click control point to delete

PaulHax committed:
feat(PointsController): new control point appears where clicked

PaulHax committed:
chore(test): cleanup

PaulHax committed:
feat(TransferFunctionEditor): more explicit entry point

PaulHax committed:
fix(ControlPoint): cleanup test with pointerup event

PaulHax committed:
feat(build): point to new entry point

PaulHax committed:
feat(examples): multi example structure

PaulHax committed:
feat(examples): added VTK.js example

PaulHax committed:
feat(Point): point promoted to class

PaulHax committed:
feat(examples): wired up VTK.js example

PaulHax committed:
feat(examples): add itk-vtk-viewer example

PaulHax committed:
Merge branch 'main' into make-transfer-function

thewtex commented on examples/itk-vtk-viewer/index.html:
```diff
@@ -0,0 +1,48 @@
+    ></div>
+    <script
+      type="text/javascript"
+      src="https://unpkg.com/itk-vtk-viewer@11.11.2/dist/itkVtkViewerCDN.js"
```
With a recent update, 11,12.1, there is a major update to the latest vtk.js, itk-wasm, and support for `unpkg.com` -> `jsdelivr.com/npm` (faster and more reliable).

PaulHax replied:
Upgraded. It does feel faster!

thewtex replied:
:+1:

PaulHax committed:
feat(examples): upgrade itk-vtk-viewer version, build fixes

thewtex reviewed (changes requested):
Getting there:

![image](https://user-images.githubusercontent.com/25432/161153234-d9f370f6-f931-4d62-8e06-e6872405830c.png)


In local testing, noticed:

- The itk-vtk-viewer example is not using this transfer function editor. This is a priority.
- The vtk-js example needs `npm i --save-dev vite`
- `npm t` results in:

```
stderr | unknown test
[Vitest] takeCoverage is not available in this NodeJs version.
Coverage could be incomplete. Update to NodeJs 14.18.

stderr | unknown test
[Vitest] takeCoverage is not available in this NodeJs version.
Coverage could be incomplete. Update to NodeJs 14.18.
```

Add .nvmrc with v16?

-

```
❯ npm run test:ui                                                                                                                                                                                                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                                                                                                                                                                                          
> itk-viewer-transfer-function-editor@0.0.1 test:ui /home/matt/src/itk-viewer-transfer-function-editor                                                                                                                                                                                                                                                                                                                                    
> vitest --ui                                                                                                                                                                                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                                                                                                                                          
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Error ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯                                                                                                                                                                                                                      
Error: The following dependencies are imported but could not be resolved:                                                                                                                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                                                                                                                                          
  itk-viewer-reference-ui/src/referenceUIMachineOptions.js (imported by /home/matt/src/itk-viewer-transfer-function-editor/examples/itk-vtk-viewer/main.js)                                                                                                                                                                                                                                                                               
  @kitware/vtk.js/favicon (imported by /home/matt/src/itk-viewer-transfer-function-editor/examples/vtk-js/VolumeMapperExample.js)                                                                                                                                                                                                                                                                                                         
  @kitware/vtk.js/macros (imported by /home/matt/src/itk-viewer-transfer-function-editor/examples/itk-vtk-viewer/main.js)                                                                                                                                                                                                                                                                                                                 
  @kitware/vtk.js/Rendering/Profiles/Volume (imported by /home/matt/src/itk-viewer-transfer-function-editor/examples/vtk-js/VolumeMapperExample.js)                                                                                                                                                                                                                                                                                       
  @kitware/vtk.js/Rendering/Core/ColorTransferFunction (imported by /home/matt/src/itk-viewer-transfer-function-editor/examples/vtk-js/VolumeMapperExample.js)                                                                                                                                                                                                                                                                            
  @kitware/vtk.js/Rendering/Misc/FullScreenRenderWindow (imported by /home/matt/src/itk-viewer-transfer-function-editor/examples/vtk-js/VolumeMapperExample.js)                                                                                                                                                                                                                                                                           
  @kitware/vtk.js/IO/Core/HttpDataSetReader (imported by /home/matt/src/itk-viewer-transfer-function-editor/examples/vtk-js/VolumeMapperExample.js)                                                                                                                                                                                                                                                                                       
  @kitware/vtk.js/Common/DataModel/PiecewiseFunction (imported by /home/matt/src/itk-viewer-transfer-function-editor/examples/vtk-js/VolumeMapperExample.js)                                                                                                                                                                                                                                                                              
  @kitware/vtk.js/Rendering/Core/Volume (imported by /home/matt/src/itk-viewer-transfer-function-editor/examples/vtk-js/VolumeMapperExample.js)                                                                                                                                                                                                                                                                                           
  @kitware/vtk.js/Rendering/Core/VolumeMapper (imported by /home/matt/src/itk-viewer-transfer-function-editor/examples/vtk-js/VolumeMapperExample.js)                                                                                                                                                                                                                                                                                     
  @kitware/vtk.js/IO/Core/DataAccessHelper/HttpDataAccessHelper (imported by /home/matt/src/itk-viewer-transfer-function-editor/examples/vtk-js/VolumeMapperExample.js)                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                                                                                                                                                          
Are they installed?                                                                                                                                                                                                                                                                                                                                                                                                                       
    at optimizeDeps (/home/matt/src/itk-viewer-transfer-function-editor/node_modules/vite/dist/node/chunks/dep-9c153816.js:55767:15)                                                                                                                                                                                                                                                                                                      
    at processTicksAndRejections (internal/process/task_queues.js:95:5)                                                                                                                                                                                                                                                                                                                                                                   
    at async runOptimize (/home/matt/src/itk-viewer-transfer-function-editor/node_modules/vite/dist/node/chunks/dep-9c153816.js:56647:44)                                                                                                                                                                                                                                                                                                 
    at async Server.httpServer.listen (/home/matt/src/itk-viewer-transfer-function-editor/node_modules/vite/dist/node/chunks/dep-9c153816.js:56662:21)                                                                                                                                                                                                                                                                                    
     ```

- `npm run lint` has warnings related to the bundles.

PaulHax committed:
fix(examples): fix lint warnings, added .nvmrc, vite to vtk.js example

PaulHax commented:
Thanks for the thorough tire kick.  Addressed the issues, except:

> The itk-vtk-viewer example is not using this transfer function editor. This is a priority.

The editor is hard to see.  I added a border.  Is that what you mean?

![image](https://user-images.githubusercontent.com/16823231/161164156-481e92fd-7190-46f4-95f7-d2030e00cfb5.png)

thewtex commented:
> . Addressed the issues,

Thank you!

> The editor is hard to see. I added a border. Is that what you mean?

Yes, I did not see it. Can it replace the existing vtkItkPiecewiseGaussianWidget instead of being placed below?

PaulHax committed:
feat(itk-vkt-viewer example): replacing vtkPiecewiseGaussianWidget

PaulHax commented:
@thewtex The vtkItkPiecewiseGaussianWidget has been completely removed from the itk-vtk-viewer example.

thewtex commented on examples/basic/main.ts:
```diff
@@ -0,0 +1,20 @@
+const editorHome = document.querySelector<HTMLDivElement>('#editorHome')
+if (editorHome) {
+  const editor = new TransferFunctionEditor(editorHome)
+  console.log('Control points', editor.getPoints())
```
debug code

thewtex commented on README.md:
````diff
@@ -1,3 +1,17 @@
+
+```
+cd examples/itk-vtk-viewer
+npm ci
````
needs a package-lock.json\

PaulHax replied:
Updated.

thewtex reviewed (approved):
@PaulHax awesome!!!

A few minor issues noted inline and good to merge!

PaulHax committed:
fix(examples): update package-lock and lint console.logs
